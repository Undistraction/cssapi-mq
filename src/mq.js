import { partial, mergeDeepLeft, merge, mergeAll } from 'ramda';

import buildRangedFeature from './features/buildRangedFeature';
import buildMediaType from './features/buildMediaType';
import buildScan from './features/buildScan';
import buildOrientation from './features/buildOrientation';

import {
  validateBreakpointSets,
  validateConfig,
  validateBreakpoints,
} from './validations';

import dimensionsOutput from './output/dimensionsOutput';
import resolutionOutput from './output/resolutionOutput';
import aspectRatioOutput from './output/aspectRatioOutput';

import { MEDIA_TYPES, UNITS } from './const';

const defaultConfig = {
  baseFontSize: 16,
  defaultMediaType: MEDIA_TYPES.SCREEN,
  dimensionsUnit: UNITS.DIMENSIONS.EM,
  shouldSeparateQueries: true,
  errorIfNoBreakpointDefined: true,
};

const configure = (breakpoints, config) => {
  const configWithDefaults = merge(defaultConfig, config);
  validateConfig(configWithDefaults);
  validateBreakpoints(breakpoints);
  validateBreakpointSets(breakpoints);
  const renderFeatures = () =>
    mergeAll([
      buildRangedFeature(
        'width',
        dimensionsOutput(config),
        breakpoints.width,
        configWithDefaults
      ),
      buildRangedFeature(
        'height',
        dimensionsOutput(config),
        breakpoints.height,
        configWithDefaults
      ),
      buildRangedFeature(
        'resolution',
        resolutionOutput(config),
        breakpoints.resolution,
        configWithDefaults
      ),
      buildRangedFeature(
        'aspect-ratio',
        aspectRatioOutput(config),
        breakpoints.aspectRatio,
        configWithDefaults
      ),
    ]);

  // ---------------------------------------------------------------------------
  // API
  // ---------------------------------------------------------------------------

  // Tweak
  // ---------------------------------------------------------------------------

  const tweak = (mq, tweakpoints) => {
    validateBreakpoints(tweakpoints);
    validateBreakpointSets(tweakpoints);
    const mergedBreakpoints = mergeDeepLeft(breakpoints, tweakpoints);
    mq.tweaked = configure(mergedBreakpoints, configWithDefaults);
    return mq;
  };

  // Media Query Features
  // ---------------------------------------------------------------------------

  // ---------------------------------------------------------------------------
  // Export
  // ---------------------------------------------------------------------------

  const exports = {
    mediaType: buildMediaType(configWithDefaults.defaultMediaType),
    orientation: buildOrientation(),
    scan: buildScan(),
    ...renderFeatures(),
  };

  exports.tweak = partial(tweak, [exports]);
  return exports;
};

export default {
  configure,
};
