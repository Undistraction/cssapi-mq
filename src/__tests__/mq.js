import './helpers/toEqualCSS';
import mq from '../mq';

// Register serializer for use by Jest in generating snapshots. Without a serializer the snapshots are difficult to read.
import cssSerialiser from './helpers/cssSerialiser';

expect.addSnapshotSerializer(cssSerialiser);

const validBreakpoints = {
  small: 400, // 0–400
  medium: 900, // 400–900
  large: 1100, // 900–1100
  xLarge: 1300, // 1100–1300
};

const validMQ = () => mq.configure(validBreakpoints);

describe('api', () => {
  describe('configuration', () => {
    it('throws if no breakpoints are supplied', () => {
      expect(() => mq.configure()).toThrow();
    });

    it('throws if invalid breakpoint value is supplied', () => {
      expect(() => mq.configure({ small: 'xxx' })).toThrow();
    });

    it("doesn't throw an error with default configuration", () => {
      expect(() => validMQ.not.toThrow());
    });

    describe('config object', () => {
      it("adjusts values based on 'basefontSize'", () => {
        const result = mq
          .configure(validBreakpoints, { baseFontSize: 10 })
          .belowWidth('small')`
        background-color: ${() => 'GhostWhite'};
      `;
        expect(result).toMatchSnapshot();
      });

      it("throws if 'baseFontSize' is not a positive number", () => {
        const config = { baseFontSize: 'xxxx' };
        expect(() => mq.configure(validBreakpoints, config)).toThrow();
      });

      it("doesn't throw an error if 'baseFontSize' is a positive number", () => {
        const config = { baseFontSize: 12 };
        expect(() => mq.configure(validBreakpoints, config)).not.toThrow();
      });

      it("throws if 'defaultMediaType' is not valid", () => {
        const config = { defaultMediaType: 'xxxx' };
        expect(() => mq.configure(validBreakpoints, config)).toThrow();
      });

      it("doesn't throw an error if 'defaultMediaType' is valid", () => {
        expect(() =>
          mq.configure(validBreakpoints, { defaultMediaType: 'all' })
        ).not.toThrow();
        // Special check for empty string
        expect(() =>
          mq.configure(validBreakpoints, { defaultMediaType: '' })
        ).not.toThrow();
      });

      it("throws if 'unit' is not valid", () => {
        const config = { unit: 'xxxx' };
        expect(() => mq.configure(validBreakpoints, config)).toThrow();
      });

      it("doesn't throw an error if 'unit' is valid", () => {
        const config = { unit: 'px' };
        expect(() => mq.configure(validBreakpoints, config)).not.toThrow();
      });

      it("throws if 'separateIfEms' is not a boolean", () => {
        const config = { separateIfEms: 'xxxx' };
        expect(() => mq.configure(validBreakpoints, config)).toThrow();
      });

      it("doesn't throw an error if 'separateIfEms' is a boolean", () => {
        const config = { separateIfEms: false };
        expect(() => mq.configure(validBreakpoints, config)).not.toThrow();
      });
    });
  });

  describe('aboveWidth', () => {
    it('returns the correct media query', () => {
      const result = validMQ().aboveWidth('small')`
        background-color: ${() => 'GhostWhite'};
      `;
      expect(result).toMatchSnapshot();
    });

    it("throws if breakpoint doesn't exist", () => {
      expect(() => validMQ().aboveWidth('xxxx')``).toThrow();
    });
  });

  describe('belowWidth', () => {
    it('returns the correct media query', () => {
      const result = validMQ().belowWidth('small')`
        background-color: ${() => 'GhostWhite'};
      `;
      expect(result).toMatchSnapshot();
    });

    it("throws if breakpoint doesn't exist", () => {
      expect(() => validMQ().belowWidth('xxxx')``).toThrow();
    });
  });

  describe('betweenWidths', () => {
    it('returns the correct media query', () => {
      const result = validMQ().betweenWidths('small', 'medium')`
        background-color: ${() => 'GhostWhite'};
      `;
      expect(result).toMatchSnapshot();
    });

    it("throws if 'from' breakpoint doesn't exist", () => {
      expect(() => validMQ().betweenWidths('xxxx', 'large')``).toThrow();
    });

    it("throws if 'to' breakpoint doesn't exist", () => {
      expect(() => validMQ().betweenWidths('large', 'xxxx')``).toThrow();
    });
  });

  describe('atBreakpointWidth', () => {
    it('returns the correct query when it is first breakpoint', () => {
      const result = validMQ().atWidthBreakpoint('small')`
        background-color: ${() => 'GhostWhite'};
      `;
      expect(result).toMatchSnapshot();
    });

    it('returns the correct query when it is last breakpoint', () => {
      const result = validMQ().atWidthBreakpoint('xLarge')`
        background-color: ${() => 'GhostWhite'};
      `;
      expect(result).toMatchSnapshot();
    });

    it('returns the correct query when it is between other breakpoints', () => {
      const result = validMQ().atWidthBreakpoint('large')`
        background-color: ${() => 'GhostWhite'};
      `;
      expect(result).toMatchSnapshot();
    });

    it("throws if breakpoint doesn't exist", () => {
      expect(() => validMQ().atWidthBreakpoint('xxxx')``).toThrow();
    });
  });
});
