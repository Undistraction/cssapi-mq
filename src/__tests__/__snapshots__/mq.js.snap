// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aspect-ratio range accessors aspectRatio() doesn't separate dimensionsUnits if not configured 1`] = `(aspect-ratio: 2/3)`;

exports[`aspect-ratio range accessors aspectRatio() returns the correct media fragment 1`] = `(aspect-ratio: 2/3)`;

exports[`aspect-ratio range accessors maxAspectRatio() doesn't separate dimensionsUnits if not configured 1`] = `(max-aspect-ratio: 2/3)`;

exports[`aspect-ratio range accessors maxAspectRatio() returns the correct media fragment 1`] = `(max-aspect-ratio: 2/3)`;

exports[`aspect-ratio range accessors minAspectRatio() doesn't separate dimensionsUnits if not configured 1`] = `(min-aspect-ratio: 2/3)`;

exports[`aspect-ratio range accessors minAspectRatio() returns the correct media fragment 1`] = `(min-aspect-ratio: 2/3)`;

exports[`aspect-ratio range queries aboveAspectRatio() returns the correct media query 1`] = `
@media screen and (min-aspect-ratio: 2/3) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`aspect-ratio range queries atAspectRatio() returns the correct media query 1`] = `
@media screen and (aspect-ratio: 2/3) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`aspect-ratio range queries atAspectRatioBreakpoint() returns the correct query when it is between other breakpoints 1`] = `
@media screen and (min-aspect-ratio: 3/2) and (max-aspect-ratio: 16/9) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`aspect-ratio range queries atAspectRatioBreakpoint() returns the correct query when it is first breakpoint 1`] = `
@media screen and (min-aspect-ratio: 2/3) and (max-aspect-ratio: 1/1) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`aspect-ratio range queries atAspectRatioBreakpoint() returns the correct query when it is last breakpoint 1`] = `
@media screen and (min-aspect-ratio: 16/9) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`aspect-ratio range queries belowAspectRatio() returns the correct media query 1`] = `
@media screen and (max-aspect-ratio: 2/3) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`aspect-ratio range queries betweenAspectRatios() returns the correct media query 1`] = `
@media screen and (min-aspect-ratio: 2/3) and (max-aspect-ratio: 1/1) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`aspect-ratio range queries betweenAspectRatios() returns the correct media query with breakpoint order reversed 1`] = `
@media screen and (min-aspect-ratio: 2/3) and (max-aspect-ratio: 1/1) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`configuration config object adjusts values based on 'basefontSize' 1`] = `
@media screen and (max-width: 39.999em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range accessors height() doesn't separate dimensionsUnits if not configured 1`] = `(height: 25em)`;

exports[`height range accessors height() renders configured dimensionsUnits 1`] = `(height: 25rem)`;

exports[`height range accessors height() renders configured dimensionsUnits 2`] = `(height: 400px)`;

exports[`height range accessors height() returns the correct media fragment 1`] = `(height: 25em)`;

exports[`height range accessors maxHeight() doesn't separate dimensionsUnits if not configured 1`] = `(max-height: 25em)`;

exports[`height range accessors maxHeight() renders configured dimensionsUnits 1`] = `(max-height: 24.999375rem)`;

exports[`height range accessors maxHeight() renders configured dimensionsUnits 2`] = `(max-height: 399px)`;

exports[`height range accessors maxHeight() returns the correct media fragment 1`] = `(max-height: 24.999375em)`;

exports[`height range accessors minHeight() doesn't separate dimensionsUnits if not configured 1`] = `(min-height: 25em)`;

exports[`height range accessors minHeight() renders configured dimensionsUnits 1`] = `(min-height: 25rem)`;

exports[`height range accessors minHeight() renders configured dimensionsUnits 2`] = `(min-height: 400px)`;

exports[`height range accessors minHeight() returns the correct media fragment 1`] = `(min-height: 25em)`;

exports[`height range queries aboveHeight() returns the correct media query 1`] = `
@media screen and (min-height: 25em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range queries atHeight() returns the correct media query 1`] = `
@media screen and (height: 25em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range queries atHeightBreakpoint() returns the correct query when it is between other breakpoints 1`] = `
@media screen and (min-height: 68.75em) and (max-height: 81.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range queries atHeightBreakpoint() returns the correct query when it is first breakpoint 1`] = `
@media screen and (min-height: 25em) and (max-height: 56.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range queries atHeightBreakpoint() returns the correct query when it is last breakpoint 1`] = `
@media screen and (min-height: 81.25em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range queries belowHeight() returns the correct media query 1`] = `
@media screen and (max-height: 24.999375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range queries betweenHeights() returns the correct media query 1`] = `
@media screen and (min-height: 25em) and (max-height: 56.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`height range queries betweenHeights() returns the correct media query with breakpoint order reversed 1`] = `
@media screen and (min-height: 25em) and (max-height: 56.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`mediaType returns the correct default media type if called with no arguments 1`] = `screen`;

exports[`mediaType returns the supplied mediaType for '' 1`] = ``;

exports[`mediaType returns the supplied mediaType for 'all' 1`] = `all`;

exports[`mediaType returns the supplied mediaType for 'print' 1`] = `print`;

exports[`mediaType returns the supplied mediaType for 'screen' 1`] = `screen`;

exports[`mediaType returns the supplied mediaType for 'speech' 1`] = `speech`;

exports[`mediaType supports multiple values 1`] = `screen, speech,`;

exports[`orientation returns the supplied orientation for 'landscape' 1`] = `(orientation: landscape)`;

exports[`orientation returns the supplied orientation for 'portrait' 1`] = `(orientation: portrait)`;

exports[`resolution range accessors maxResolution() doesn't separate dimensionsUnits if not configured 1`] = `(max-resolution: 72dpi)`;

exports[`resolution range accessors maxResolution() returns the correct media fragment 1`] = `(max-resolution: 71dpi)`;

exports[`resolution range accessors minResolution() doesn't separate dimensionsUnits if not configured 1`] = `(min-resolution: 72dpi)`;

exports[`resolution range accessors minResolution() returns the correct media fragment 1`] = `(min-resolution: 72dpi)`;

exports[`resolution range accessors resolution() doesn't separate dimensionsUnits if not configured 1`] = `(resolution: 72dpi)`;

exports[`resolution range accessors resolution() returns the correct media fragment 1`] = `(resolution: 72dpi)`;

exports[`resolution range queries aboveResolution() returns the correct media query 1`] = `
@media screen and (min-resolution: 72dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`resolution range queries atResolution() returns the correct media query 1`] = `
@media screen and (resolution: 72dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`resolution range queries atResolutionBreakpoint() returns the correct query when it is between other breakpoints 1`] = `
@media screen and (min-resolution: 300dpi) and (max-resolution: 599dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`resolution range queries atResolutionBreakpoint() returns the correct query when it is first breakpoint 1`] = `
@media screen and (min-resolution: 72dpi) and (max-resolution: 149dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`resolution range queries atResolutionBreakpoint() returns the correct query when it is last breakpoint 1`] = `
@media screen and (min-resolution: 600dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`resolution range queries belowResolution() returns the correct media query 1`] = `
@media screen and (max-resolution: 71dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`resolution range queries betweenResolutions() returns the correct media query 1`] = `
@media screen and (min-resolution: 72dpi) and (max-resolution: 149dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`resolution range queries betweenResolutions() returns the correct media query with breakpoint order reversed 1`] = `
@media screen and (min-resolution: 72dpi) and (max-resolution: 149dpi) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`scan returns the supplied scan for 'interlace' 1`] = `(scan: interlace)`;

exports[`scan returns the supplied scan for 'progressive' 1`] = `(scan: progressive)`;

exports[`tweaked adds includes both old breakpoints and added tweakpoints 1`] = `
@media screen and (min-width: 18.75em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`tweaked adds includes both old breakpoints and added tweakpoints 2`] = `
@media screen and (min-width: 68.75em) and (max-width: 18.749375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`tweaked doesn't effect the original mq 1`] = `
@media screen and (min-width: 25em) and (max-width: 56.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range accessors maxWidth() doesn't separate dimensionsUnits if not configured 1`] = `(max-width: 25em)`;

exports[`width range accessors maxWidth() renders configured dimensionsUnits 1`] = `(max-width: 24.999375rem)`;

exports[`width range accessors maxWidth() renders configured dimensionsUnits 2`] = `(max-width: 399px)`;

exports[`width range accessors maxWidth() returns the correct media fragment 1`] = `(max-width: 24.999375em)`;

exports[`width range accessors minWidth() doesn't separate dimensionsUnits if not configured 1`] = `(min-width: 25em)`;

exports[`width range accessors minWidth() renders configured dimensionsUnits 1`] = `(min-width: 25rem)`;

exports[`width range accessors minWidth() renders configured dimensionsUnits 2`] = `(min-width: 400px)`;

exports[`width range accessors minWidth() returns the correct media fragment 1`] = `(min-width: 25em)`;

exports[`width range accessors width() doesn't separate dimensionsUnits if not configured 1`] = `(width: 25em)`;

exports[`width range accessors width() renders configured dimensionsUnits 1`] = `(width: 25rem)`;

exports[`width range accessors width() renders configured dimensionsUnits 2`] = `(width: 400px)`;

exports[`width range accessors width() returns the correct media fragment 1`] = `(width: 25em)`;

exports[`width range queries aboveWidth() returns the correct media query 1`] = `
@media screen and (min-width: 25em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range queries atWidth() returns the correct media query 1`] = `
@media screen and (width: 25em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range queries atWidthBreakpoint() returns the correct query when it is between other breakpoints 1`] = `
@media screen and (min-width: 68.75em) and (max-width: 81.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range queries atWidthBreakpoint() returns the correct query when it is first breakpoint 1`] = `
@media screen and (min-width: 25em) and (max-width: 56.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range queries atWidthBreakpoint() returns the correct query when it is last breakpoint 1`] = `
@media screen and (min-width: 81.25em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range queries belowWidth() returns the correct media query 1`] = `
@media screen and (max-width: 24.999375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range queries betweenWidths() returns the correct media query 1`] = `
@media screen and (min-width: 25em) and (max-width: 56.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;

exports[`width range queries betweenWidths() returns the correct media query with breakpoint order reversed 1`] = `
@media screen and (min-width: 25em) and (max-width: 56.249375em) {
    background-color: function () {
        return 'GhostWhite';
    }; ;
}
`;
